mllib/src/main/scala/org/apache/spark/ml/feature/OneHotEncoderEstimator.scala

```diff
@@ -0,0 +1,464 @@
+    val outputColNames = $(outputCols)
+    val inputFields = schema.fields
+
+    require(inputColNames.length == outputColNames.length,
+      s"The number of input columns ${inputColNames.length} must be the same as the number of " +
+        s"output columns ${outputColNames.length}.")
+
+    val outputFields = inputColNames.zip(outputColNames).map { case (inputColName, outputColName) =>
+
+      require(schema(inputColName).dataType.isInstanceOf[NumericType],
+        s"Input column must be of type NumericType but got ${schema(inputColName).dataType}")
+      require(!inputFields.exists(_.name == outputColName),
+        s"Output column $outputColName already exists.")
+
+      OneHotEncoderCommon.transformOutputColumnSchema(
+        schema(inputColName), $(dropLast), outputColName)
+    }
+    StructType(inputFields ++ outputFields)
+  }
+
+  @Since("2.3.0")
+  override def fit(dataset: Dataset[_]): OneHotEncoderModel = {
+    val transformedSchema = transformSchema(dataset.schema)
+    val categorySizes = new Array[Int]($(outputCols).length)
+
+    val columnToScanIndices = $(outputCols).zipWithIndex.flatMap { case (outputColName, idx) =>
+      val numOfAttrs = AttributeGroup.fromStructField(
+        transformedSchema(outputColName)).size
+      if (numOfAttrs < 0) {
+        Some(idx)
+      } else {
+        categorySizes(idx) = numOfAttrs
+        None
+      }
+    }
+
+    // Some input columns don't have attributes or their attributes don't have necessary info.
+    // We need to scan the data to get the number of values for each column.
+    if (columnToScanIndices.length > 0) {
+      val inputColNames = columnToScanIndices.map($(inputCols)(_))
+      val outputColNames = columnToScanIndices.map($(outputCols)(_))
+      val attrGroups = OneHotEncoderCommon.getOutputAttrGroupFromData(
+        dataset, $(dropLast), inputColNames, outputColNames)
+      attrGroups.zip(columnToScanIndices).foreach { case (attrGroup, idx) =>
+        categorySizes(idx) = attrGroup.size
+      }
+    }
+
+    val model = new OneHotEncoderModel(uid, categorySizes).setParent(this)
+    copyValues(model)
+  }
+
+  @Since("2.3.0")
+  override def copy(extra: ParamMap): OneHotEncoderEstimator = defaultCopy(extra)
+}
+
+@Since("2.3.0")
+object OneHotEncoderEstimator extends DefaultParamsReadable[OneHotEncoderEstimator] {
+
+  private[feature] val SKIP_INVALID: String = "skip"
+  private[feature] val ERROR_INVALID: String = "error"
+  private[feature] val supportedHandleInvalids: Array[String] = Array(SKIP_INVALID, ERROR_INVALID)
+
+  @Since("2.3.0")
+  override def load(path: String): OneHotEncoderEstimator = super.load(path)
+}
+
+@Since("2.3.0")
+class OneHotEncoderModel private[ml] (
+    @Since("2.3.0") override val uid: String,
+    @Since("2.3.0") val categorySizes: Array[Int])
+  extends Model[OneHotEncoderModel] with OneHotEncoderParams with MLWritable {
+
+  import OneHotEncoderModel._
+
+  private def encoders: Array[UserDefinedFunction] = {
```

Comment:
What about defining one UDF `(label, size) => Vector` instead of defining an array of UDFs ?
Or do you find some apparent perf difference between them ?
If not, I prefer defining only one UDF, which makes the code clearer.

Reply:
Sounds good. Use one UDF now.